### PR TITLE
[stable] Fix building on Musl

### PR DESCRIPTION
**Pull request by Geod24**
This helps with building DMD on Musl. I am currently working on [a package for Alpine Linux](https://gitlab.alpinelinux.org/alpine/aports/merge_requests/2880) and needed a few tweaks to get things to work.
Targeting stable as the fixes are minor.

### CODE REVIEW

File: src/dmd/e2ir.d [hunk 1]
```diff
@@ -6323,17 +6334,17 @@ elem *callCAssert(IRState *irs, const ref Loc loc, Expression exp, Expression em
...
-            id = fd.toPrettyChars();
-        const len = strlen(id);
-        Symbol *si = toStringSymbol(id, len, 1);
-        elem *efunc = el_ptr(si);
-
+        elem* efunc = getFuncName();
         auto eassert = el_var(getRtlsym(RTLSYM_C__ASSERT_RTN));
         ea = el_bin(OPcall, TYvoid, eassert, el_params(elmsg, eline, efilename, efunc, null));
     }
+    else version (CRuntime_Musl)
```
Comment (thewilsonator) [inline src/dmd/e2ir.d]:
```
if(...) {...}
else version(...) {...}
else {...}
```
looks _really_ weird...

File: src/dmd/mars.d [hunk 2]
```diff
@@ -1188,7 +1188,14 @@ void addDefaultVersionIdentifiers(const ref Param params)
...
         VersionCondition.addPredefinedGlobalIdent("ELFv1");
-        VersionCondition.addPredefinedGlobalIdent("CRuntime_Glibc");
+        // Note: This should be done with a triplet, to support cross compilation
+        // However DMD currently does not support it, so this is a simple
+        // fix to make DMD compile on Musl-based systems such as Alpine.
+        // See https://github.com/dlang/dmd/pull/8020
+        version (CRuntime_Musl)
+            VersionCondition.addPredefinedGlobalIdent("CRuntime_Musl");
+        else
+            VersionCondition.addPredefinedGlobalIdent("CRuntime_Glibc");
```
Comment (thewilsonator) [inline src/dmd/mars.d]: ... even with this.

File: src/dmd/e2ir.d (hunk 1 above)
Comment (Geod24) [reply src/dmd/e2ir.d]: It does, but it actually is quite natural when you think of it. After all:
```D
if (a) {}
else
{
    if (b) {}
    else {}
}
```
is equivalent to:
```D
if (a) {}
else if (b) {}
else {}
```

And I suppose the following would look less weird:
```D
if (a) {}
else
{
    version (b) {}
    else {}
}
```

but works just the same.

File: src/dmd/mars.d (hunk 2 above)
Comment (Geod24) [reply src/dmd/mars.d]: What do you mean ?

Comment (thewilsonator) [reply src/dmd/mars.d]: I mean I get why you had the version in the if above, i.e. because if this.

**Comment by thewilsonator** [comment]
Sorry if that wasn't obvious, using
```
if (a) {}
else
{
    version (b) {}
    else {}
}
```
that form would be appreciated.

File: src/dmd/mars.d [hunk 3]
```diff
@@ -1188,7 +1188,14 @@ void addDefaultVersionIdentifiers(const ref Param params)
         VersionCondition.addPredefinedGlobalIdent("Posix");
         VersionCondition.addPredefinedGlobalIdent("linux");
         VersionCondition.addPredefinedGlobalIdent("ELFv1");
-        VersionCondition.addPredefinedGlobalIdent("CRuntime_Glibc");
+        // Note: This should be done with a triplet, to support cross compilation
```
Comment (WalterBright) [inline src/dmd/mars.d]: What's a "triplet"?

Comment (Geod24) [reply src/dmd/mars.d]: https://wiki.osdev.org/Target_Triplet

**Review by WalterBright — APPROVED** [review]
This PR looks unprincipled, there should be some sort of principle around these choices. I fear an impossible tangle is being created (that mixed if/version statement is suspect). But I'll pass on it anyway as I have no idea what Musl is or what its peculiar problems are.

File: src/dmd/mars.d (hunk 3 above)
Comment (WalterBright) [reply src/dmd/mars.d]: Please put that in the code comment, I doubt I'm the only one wondering what the comment is talking about.

**Comment by PetarKirov** [comment]
@WalterBright [Musl](https://www.musl-libc.org/) is an alternative libc implementation for Linux (see [comparison](http://www.etalabs.net/compare_libcs.html) for extra information). One of the reasons for its huge appeal in recent years is the [Alpine Linux](https://en.m.wikipedia.org/wiki/Alpine_Linux) distribution, which is very popular base for containerized [12 factor apps](https://12factor.net/).

**Comment by WalterBright** [comment]
@PetarKirov thanks for the illumination.

**Comment by Geod24** [comment]
Rebased on stable and addressed comments.

**Comment by Geod24** [comment]
Green.

**Pull request merged**